### PR TITLE
Example Horizontal Outerloop Control Module

### PR DESCRIPTION
**Pull request by dewagter**
Many projects have an own outer loop controller that takes the state and sets the angles. This forms a quick start for them.

### CODE REVIEW

File: sw/airborne/modules/ctrl/ctrl_module_outerloop_demo.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  ctrl.rc_y = 0;
+}
+
+void guidance_h_module_read_rc(void)
+{
+  // -MAX_PPRZ to MAX_PPRZ
+  //ctrl.rc_t = radio_control.values[RADIO_THROTTLE];
+  ctrl.rc_x = radio_control.values[RADIO_ROLL];
+  ctrl.rc_y = radio_control.values[RADIO_PITCH];
+  //ctrl.rc_z = radio_control.values[RADIO_YAW];
```
Comment (kirkscheper) [inline sw/airborne/modules/ctrl/ctrl_module_outerloop_demo.c]: Why are these commented out? The values are not used in this example but doesn't mean they shouldn't be set.

**Comment by noether** [comment]
Ewoud and I wrote an ABI message to command accelerations in the navigation frame to INDI.
Maybe you would like to add such a message in this example? (or I can do it afterward).

I agree with @gautierhattenberger about the autopilot code generation. However, it is true that sometimes you only want to command attitude angles or velocities from a single "separated/isolated" outer loop. So this example always sounds more convenient (especially for those who are new to Paparazzi like new students).

**Comment by dewagter** [comment]
@noether Setting accelerations sound great. In which branch did you do that?
@gautierhattenberger I like the idea of autopilot generation, but is there a better way than copy pasting the existing rotorcraft_autopilot.xml and changing? Most would remain the same and this is not simple to maintain.

The goal is to have an airframe or module in which people can set their own desired accelerations, e.g. form a companion computer board:

body_accel = own_controller(state);

**Comment by kirkscheper** [comment]
It's in master, code in https://github.com/paparazzi/paparazzi/blob/master/sw/airborne/firmwares/rotorcraft/guidance/guidance_indi.c. You just have to write to the ACCEL_SP abi.

File: sw/airborne/modules/ctrl/ctrl_module_outerloop_demo.c [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+void guidance_h_module_run(bool in_flight)
+{
+  // Demo which copies the sticks into pitch and roll
+  float roll = ctrl.rc_x;
+  roll /= (float) MAX_PPRZ;
+  roll *= RadOfDeg(10);
+
+  float pitch = ctrl.rc_y;
+  pitch /= (float) MAX_PPRZ;
+  pitch *= RadOfDeg(10);
```
[the remark this answers is not in the document]
Comment (dewagter) [reply sw/airborne/modules/ctrl/ctrl_module_outerloop_demo.c]: The goal is to replace this code and set your own. This code is merely to test the empty module. I could, however remove this code altogether.

**Comment by dewagter** [comment]
Acceleration setpoint in INDI is only working well when good speed / position measurements are available. This module is typically used with only vision.

**Comment by gautierhattenberger** [comment]
Indeed, if you use AP generation, you need to provide a template that student can modify. It is not different than providing a Simulink diagram, with the same issues for code duplication I guess.
If you think working with this example is easier for you at the moment, I don't mind to merge it.

**Comment by noether** [comment]
@dewagter 

The acceleration setpoint in INDI does not depend on speed/position. How good the vehicle tracks the desired acceleration will depend on the physical model (thrust, delay of the actuators, inertia..), which is needed by the INDI. It is true that the North and Up (navigation frame) are needed, and those are easy to estimate if we are in a controlled environment (vision). 

Maybe we confused you because right now the acceleration commands only work when the vehicle is on "guidance_indi" mode, which needs speed and position (but when you send the acc commands, they are not used at all). Indeed, maybe it is not the best place to place our feature. In fact, if your module stops sending the acceleration commands, the vehicle will resume the current block of the flight plan and go to the current waypoint.

**Comment by dewagter** [comment]
- removed commented code
 - added descriptive comments
 - fixed doxygen
 - Added suggestion about ABI setpoint so people know about it

**Pull request merged**